vowpalwabbit/options.h

```diff
@@ -121,23 +121,51 @@ struct typed_option : base_option
   typed_option& value(T value)
   {
     m_value = std::make_shared<T>(value);
+    value_set_callback(value);
     return *this;
   }
 
   T value() const { return m_value ? *m_value : T(); }
 
-  T& m_location;
+protected:
+  // Allows inheriting classes to handle set values. Noop by default.
+  virtual void value_set_callback(const T& /*value*/) {}
 
 private:
   // Would prefer to use std::optional (C++17) here but we are targeting C++11
   std::shared_ptr<T> m_value{nullptr};
   std::shared_ptr<T> m_default_value{nullptr};
 };
 
+// The contract of typed_option_with_location is that the first set of the option value is written to the given
+// location, otherwise it is a noop.
 template <typename T>
-option_builder<typed_option<T>> make_option(std::string name, T& location)
+struct typed_option_with_location : typed_option<T>
 {
-  return option_builder<typed_option<T>>(name, location);
+  typed_option_with_location(const std::string& name, T& location) : typed_option<T>(name), m_location{&location} {}
+  virtual void value_set_callback(const T& value) override
+  {
+    if (m_location != nullptr)
+    {
+      *m_location = value;
+      m_location = nullptr;

```

Comment:
why are we setting as nullptr?

Reply:
Since the location often points to a stack local you can only safely set the location in the first add and parse call. By setting it to null here we say only the first set gets written to location. 

However, I can see this being a problem when we set the value of an option before the add and parse call (such as when insert is replaced with set). Honestly, the whole location to write to thing is rather unsafe and makes the options framework more complicated.

Reply:
Reworked it so that add and parse signals through to here instead